##### Libraries

In [106]:
from matplotlib.pyplot import *
from numpy import *
from pandas import read_csv,DataFrame,Series

##### Data

In [ ]:
rho = 0.997
empty = {
    #Container,Expected_Volume(ml),Weight(g)
    "MC"   :[[0,mean([33.4249,33.4239,33.4240])]], #Measuring Cylinder
    "VF"   :[[0,mean([66.2980,66.2970,66.2980])]], #Volumetric Flask
    "Bk"   :[[0,mean([28.4140,28.4130,28.4130])]], #Beaker
    "NP"   :[[0,mean([28.4140,28.4130,28.4130])]], #Same as Beaker
    "GP"   :[[0,mean([28.4140,28.4130,28.4130])]], #Same as Beaker
    "Bk+TT":[[0,mean([40.5660,40.5650,40.5650])]], #Test-tube in beaker
}

full = {
    "GP"   :[
        [10,mean([38.3850,38.3860,38.3870])],
        [10,38.3317],
        [10,38.4670],
        ],
    "NP"   :[
        [10,mean([38.3780,38.3770,38.3760])],
        [10,mean([38.283,38.384,38.384])],
        [10,38.377],
        ],
    "VF"   :[
        [100,mean([165.543,165.541,165.539])],
        [100,mean([165.695,165.693,165.699])],
        [100,mean([165.477,165.476,165.477])],
        ],
    "MC"   :[
        [10,mean([43.523,43.525,43.523])],
        [10,43.4600],
        [10,43.4020],
    ],
    "Bk+TT":[
        [4,44.702],
        [4,44.783],
        [4,44.614],
    ],
}

##### Volume (mL) vs weights (g)

In [ ]:
water = {t : [x[:-1] + [round(x[-1]-empty[t][0][1],4)] for x in full[t]] for t in full}
for t in water:
    print(t,water[t])

GP [[10, 9.9727], [10, 9.9184], [10, 10.0537]]
NP [[10, 9.9637], [10, 9.937], [10, 9.9637]]
VF [[100, 99.2433], [100, 99.398], [100, 99.179]]
MC [[10, 10.0994], [10, 10.0357], [10, 9.9777]]
Bk+TT [[4, 4.1367], [4, 4.2177], [4, 4.0487]]


In [ ]:
volume0 = {}
volume1 = {}
volume2 = {}
volume = {}
for t in water:
    volume0[t] = water[t][0][1]/rho
    volume1[t] = water[t][1][1]/rho
    volume2[t] = water[t][2][1]/rho
    volume[t] = water[t][0][0]
volume0 = Series(volume0)
volume1 = Series(volume1)
volume2 = Series(volume2)
volume = Series(volume)

##### Calculated Weights vs measured weights (g)

In [ ]:
for t in water:
    for x in water[t]:
        x[0] = round(x[0]*rho,4)
for t in water:
    print(t,water[t])
calc_weights = {t:water[t][0][0] for t in water}

GP [[9.97, 9.9727], [9.97, 9.9184], [9.97, 10.0537]]
NP [[9.97, 9.9637], [9.97, 9.937], [9.97, 9.9637]]
VF [[99.7, 99.2433], [99.7, 99.398], [99.7, 99.179]]
MC [[9.97, 10.0994], [9.97, 10.0357], [9.97, 9.9777]]
Bk+TT [[3.988, 4.1367], [3.988, 4.2177], [3.988, 4.0487]]


##### Errors (in g)

In [ ]:
for t in water:
    for i in range(len(water[t])):
        water[t][i] = round(water[t][i][1] - water[t][i][0],3)
X = array(list(water.values()))
print(X)
df = DataFrame(X,index = water.keys(),columns = [0,1,2])
for t in water:
    print(t,water[t])
df

[[ 0.0027 -0.0516  0.0837]
 [-0.0063 -0.033  -0.0063]
 [-0.4567 -0.302  -0.521 ]
 [ 0.1294  0.0657  0.0077]
 [ 0.1487  0.2297  0.0607]]
GP [0.0027, -0.0516, 0.0837]
NP [-0.0063, -0.033, -0.0063]
VF [-0.4567, -0.302, -0.521]
MC [0.1294, 0.0657, 0.0077]
Bk+TT [0.1487, 0.2297, 0.0607]


,0,1,2
GP,0.0027,-0.0516,0.0837
NP,-0.0063,-0.0330,-0.0063
VF,-0.4567,-0.3020,-0.5210
MC,0.1294,0.0657,0.0077
Bk+TT,0.1487,0.2297,0.0607


##### RMSE (g)

In [ ]:
rmse = {}
for t in water:
    rmse[t] = round(mean(array(water[t])**2)**0.5,3)
for t in water:
    print(t,round(rmse[t],3))
print("Total :",round(mean(array(list(rmse.values()))**2)**0.5,3))
df["RMSE"] = Series(rmse)
df["Mean Deviation"] = (df[0] + df[1] + df[2])/3
df["Mean Absolute deviation"] = (abs(df[0]) + abs(df[1]) + abs(df[2]))/3
df["V0"] = volume0
df["V1"] = volume1
df["V2"] = volume2
df["Mean Volume calculated from weight"] = (volume0 + volume1 + volume2)/3
df["desired volume"] = volume
df["calculated weight"] = Series(calc_weight)
df = round(df,3)
df

GP 0.057
NP 0.02
VF 0.436
MC 0.084
Bk+TT 0.162
Total : 0.213


,0,1,2,RMSE,Mean Deviation,Mean Absolute deviation,V0,V1,V2,Mean Volume calculated from weight,desired volume,maxium deviation
GP,0.003,-0.052,0.084,0.057,0.012,0.046,10.003,9.948,10.084,10.012,10,0.084
NP,-0.006,-0.033,-0.006,0.020,-0.015,0.015,9.994,9.967,9.994,9.985,10,0.033
VF,-0.457,-0.302,-0.521,0.436,-0.427,0.427,99.542,99.697,99.477,99.572,100,0.521
MC,0.129,0.066,0.008,0.084,0.068,0.068,10.130,10.066,10.008,10.068,10,0.129
Bk+TT,0.149,0.230,0.061,0.162,0.146,0.146,4.149,4.230,4.061,4.147,4,0.230


##### Deviation of calculated value from average measured value

In [ ]:
S = 0
for t in water:
    x = mean(water[t])
    print(t,round(x,3))
    S += abs(x)
S /= 5
print("mean deviation",round(S,3))

GP 0.012
NP -0.015
VF -0.427
MC 0.068
Bk+TT 0.146
mean deviation 0.133


##### Worst case scenario

In [ ]:
S = 0
maxdev = {}
for t in water:
    x = max(abs(array(water[t])))
    print(t,round(x,3))
    maxdev[t] = x
    S  = max(x,S)
print("Combined :",S)
df["maxium deviation"] = maxdev

GP 0.084
NP 0.033
VF 0.521
MC 0.129
Bk+TT 0.23
Combined : 0.521


Since the LC is 0.0001 g , so the LC error is insignificant in comparison to the human errors. 

In [ ]:
df.to_csv("Table.csv")

In [ ]:
mean(df["Mean Absolute deviation"])

0.14034666666666668

In [ ]:
read_csv("Table.csv")

,Unnamed: 0,0,1,2,RMSE,Mean Deviation,Mean Absolute deviation,V0,V1,V2,Mean Volume calculated from weight,desired volume,maxium deviation
0,GP,0.0027,-0.0516,0.0837,0.0568,0.011600,0.046000,10.002708,9.948245,10.083952,10.011635,10,0.0837
1,NP,-0.0063,-0.0330,-0.0063,0.0197,-0.015200,0.015200,9.993681,9.966901,9.993681,9.984754,10,0.0330
2,VF,-0.4567,-0.3020,-0.5210,0.4364,-0.426567,0.426567,99.541926,99.697091,99.477432,99.572150,100,0.5210
3,MC,0.1294,0.0657,0.0077,0.0839,0.067600,0.067600,10.129789,10.065898,10.007723,10.067803,10,0.1294
4,Bk+TT,0.1487,0.2297,0.0607,0.1618,0.146367,0.146367,4.149147,4.230391,4.060883,4.146807,4,0.2297


For all the containers, the maximum deviation is 0.0521 , mean deviation is 0.133, mean absolute deviation is 0.1403 , rmse is 0.213